[PLUGIN-1333] Kafka Sink Performance Improvement

katukota-google opened the pull request:
1. Make producer.send() function asynchronous
2. Set property linger.ms to 100ms which will make the producer wait for 100ms so that messages can be sent out as a batch.

katukota-google committed:
Kafka Sink Performance Improvement
1. Make producer.send() function asynchronous
2. Set property linger.ms to 100ms which will make the producer wait for 100ms so that messages can be sent out as a batch.

fernst commented on kafka-plugins-client/src/main/java/io/cdap/plugin/sink/KafkaOutputFormat.java:
```diff
@@ -43,85 +43,84 @@ public class KafkaOutputFormat extends OutputFormat<Text, Text> {
-    // Configure the properties for kafka.
-    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,
+    // CDAP-9178: cached the producer object to avoid being created on every batch interval
+    if (producer == null) {
```
nit: it would be nice if this logic to initialize the producer is moved to a separate function for cleanliness

katukota-google replied:
Fixed.

tivv commented:
Please file a JIRA and refer to it in the PR. This helps informing users on the changes

greeshmaswaminathan commented on kafka-plugins-client/src/main/java/io/cdap/plugin/sink/KafkaOutputFormat.java:
```diff
@@ -43,85 +43,84 @@ public class KafkaOutputFormat extends OutputFormat<Text, Text> {
               configuration.get(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG));
-    props.put("compression.type", configuration.get("compression.type"));
+      props.put("compression.type", configuration.get("compression.type"));
+      props.put(ProducerConfig.LINGER_MS_CONFIG, 100);
```
Any particular reason for the value 100?

katukota-google replied:
There is a white paper by Confluence: https://www.confluent.io/blog/optimizing-apache-kafka-deployment/ that suggests linger.ms values should be between 10-100 ms for best performance. I tried different values 10,50,100 and noticed that it performed best with the value of 100 ms.

The number of records generated per 30 mins for different values for pipeline configuration DataGenerator -> Wrangler -> KafkaSink were:
linger.ms 10ms- 133,243,283
linger.ms 50ms- 135,367,345
linger.ms 100ms- 142,434,822

snehalakshmisha replied:
You can add a comment in the code explaining why you are using the value 100

tivv replied:
We should allow to configure it either from UI or from runtime arguments

greeshmaswaminathan replied:
+1 to making it configurable. UI allows `Additional Kafka Producer Properties`, so maybe we can look for the property value in configuration and set it to a default value if not specified.

katukota-google replied:
Fixed.

katukota-google committed:
Kafka Sink Performance Improvement
1. Make producer.send() function asynchronous
2. Set property linger.ms to a default value of 100ms which will make the producer wait for 100ms so that messages can be sent out as a batch.

greeshmaswaminathan reviewed (approved):
Squash the commits before merge
